srv/modules/runners/openstack.py

```diff
@@ -0,0 +1,212 @@
+                    failed.append(state)
+        else:
+            # This could happen if the SLS being applied somehow doesn't exist,
+            # e.g. "No matching sls found for 'ceph.openstack' in env 'base'".
+            # Realistically this should never /actually/ happen.
+            failed.append(states)
+    if failed:
+        return {'ERROR': failed}
+
+    runner = salt.runner.RunnerClient(__opts__)
+
+    # Set up prefix for subsequent string concatenation to match what's done
+    # in the SLS files for keyring and pool names.
+    prefix = ""
+    if "prefix" in kwargs:
+        prefix = kwargs['prefix'] + "-"
+
+    conf = configparser.RawConfigParser()
+    with open("/srv/salt/ceph/configuration/cache/ceph.conf") as lines:
+        conf.read_string('\n'.join(line.strip() for line in lines))
+
+    radosgw_url = ""
+
+    rgw_urls = []
+    rgw_configurations = runner.cmd('select.from',
+                                    ['pillar=rgw_configurations', 'role=rgw', 'attr=host'],
+                                    print_event=False)
+    for rgw in rgw_configurations:
+        section = "client." + rgw[0] + "." + rgw[1]
+        if not conf.has_section(section):
+            continue
+        if conf.has_option(section, "rgw frontends") and conf.has_option(section, "rgw dns name"):
+            https = re.match(r'.*port=(\d+)s.*', conf.get(section, "rgw frontends"))
+            http = re.match(r'.*port=(\d+).*', conf.get(section, "rgw frontends"))
+            if not http and not https:
+                continue
+            if http:
+                url = "http://" + conf.get(section, "rgw dns name") + ":" + http.group(1)
+            if https:
+                url = "https://" + conf.get(section, "rgw dns name") + ":" + https.group(1)
+            url += "/swift/v1"
+            rgw_urls.append(url)
+    if len(rgw_urls) == 1:
+        radosgw_url = rgw_urls[0]
+    elif len(rgw_urls) > 1:
+        # TODO: This is...  Not ideal.  But I don't have any better ideas right now.
+        radosgw_url = "FIXME: " + " ".join(rgw_urls)
+
+    return {
+        'ceph_conf': {
+            'fsid': list(local.cmd(master_minion, 'pillar.get', ['fsid']).items())[0][1],
+            'mon_initial_members': ", ".join(runner.cmd('select.minions',
+                                             ['cluster=ceph', 'roles=mon', 'host=True'],
+                                             print_event=False)),
+            'mon_host': ", ".join(runner.cmd('select.public_addresses',
+                                             ['cluster=ceph', 'roles=mon'], print_event=False)),
+            'public_network': list(local.cmd(master_minion, 'pillar.get',
+                                             ['public_network']).items())[0][1],
+            'cluster_network': list(local.cmd(master_minion, 'pillar.get',
+                                              ['cluster_network']).items())[0][1]
```

Comment:
I could see a lot of 'list index out of rage' errors when something before those calls fails. 

Maybe we should try to gather the pillar data before and validate it somehow. ( lots of somehows and maybes in there )

Reply:
:-)